skills/repo.go

```diff
@@ -0,0 +1,96 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package skills
+
+import (
+	"errors"
+	"sort"
+	"sync"
+)
+
+// Repository stores all known snappy skills and slots and types.
+type Repository struct {
+	// Protects the internals from concurrent access.
+	m     sync.Mutex
+	types []Type
+}
+
+var (
+	// ErrDuplicate is reported when type, skill or slot already exist.
+	ErrDuplicate = errors.New("duplicate found")
```

Comment:
Can we please default to the other side?  Things should be an inline fmt.Errorf until we have a good reason not to do that. They should also include a helpful message (what is duplicated?).